Library/MeasurePlugin.h

```diff
@@ -23,6 +23,7 @@ typedef void (*NEWRELOAD)(void*, void*, double*);
 typedef void (*NEWFINALIZE)(void*);
 typedef double (*NEWUPDATE)(void*);
 typedef LPCWSTR (*NEWGETSTRING)(void*);
+typedef LPCWSTR (*GETIMAGE)(void*, PluginImageData&);

```

Comment:
Is this a good function signature? i was considering a `void` return type but i choosed to go with that to keep it similar to `GetString`.

Reply:
It is not.
The function signature must be compatible with custom functions, so when anybody writes `[MyMeasure:GetImage()]` nothing breaks. And when we try to get an image from an old plugin that has `GetImage` as a section variable, then nothing breaks.
It's actually a huge pain in the ass to come up with something simple yet reliable for this.
I'm not yet sure what to do with this.

Reply:
Kinda easy way to do thins would be to add a callback to rainmeter, to indicate that this plugin supports images.
Like, you write `RmSetImagesSupport(true)`, and rainmeter knows it can call the `GetImage` function with different signature.
The issue I have with this solution is that my LocalPluginLoader will break :)
Also, this would require not only changes in plugin api, but also in rainmeter plugin api, which decreases the changes that rainmeter devs will like this. Well, I think. Maybe they would actually like this more. Maybe we could ask them after we make a PR into upstream repository, if they like this.

Reply:
>Another solution which makes a lot of sense to me is just let the skin authors set the minimum rainmeter to run there skin to a higher version than the version that will support this feature initially, which is also a good thing because users will update there rainmeters (except if something broke there skin setup or layout and they regretted it).

Hmm.
Actually, if we add a new function into rainmeter plugin API, and a plugin uses this function, then this will one-way break compatibility.
So old plugins will work in either version, and the new plugins will only work in version where this new function exists.
So if plugin uses in-memory image transfer with new function signature, then it will just not work with old rainmeter versions.
Unfortunately, users won't get any meaningful error message about why the plugin doesn't work, but we could just say in the documentation that plugin only works with new versions, and users will likely figure this out.

I didn't test anything, so maybe I'm wrong. But your could test this.

Another issue that arises here is _what_ should be added into rainmeter api. Because, for example, it would be very convenient to _at lest_ have the `RmGetVersion`! And maybe something else. Like, `RmIsMemoryImagesEnabled` to not break my LocalPluginLoader plugin :)

Reply:
>It's actually a huge pain in the ass to come up with something simple yet reliable for this.
>I'm not yet sure what to do with this.

~~My first attempts was using the holy `void *`, but i misunderstood you before and thought the function should be similar to the current built in functions.~~

nvm, even that can be called. But i thought about an ugly solution, renaming it to `RmGetImage`, i bet that no plugin uses this with the new signature (without `PluginImageData` argument).

>The issue I have with this solution is that my LocalPluginLoader will break :)

plz don't scare me, i prefer living without this feature on breaking that plugin lol, i completely rely on developing my plugins using it.

>Well, I think. Maybe they would actually like this more. Maybe we could ask them after we make a PR into upstream repository, if they like this more.

Of course they would ask for changes, and i'm ready to refactor the entire implementation, it's just i hope they don't ask for something out of my knowledge (i.e. `RmSetImagesSupport`, i had no idea how to do this \`_\`).

Reply:
>Unfortunately, users won't get any meaningful error message about why the plugin doesn't work, but we could just say in the documentation that plugin only works with new versions, and users will likely figure this out.

"This plugin requires Rainmeter v4.x.x or higher..."

>I didn't test anything, so maybe I'm wrong. But your could test this.

Yep i will, my idea was to do this check in `init` function then log that message, and another thing, it might be annoying but since we can call bangs inside plugins, we can open the logs windows so we can be sure they have seen the message, and from there it's up to them.

Personally i don't update rainmeter except if they added a cool feature or a cool skin requires a higher version.

>Like, RmIsMemoryImagesEnabled to not break my LocalPluginLoader plugin :)

I told you i can't live without this plugin😂 i'll figure it out, <sub>hopefully😿</sub>.

Reply:
> `RmGetImage`, i bet that no plugin uses this

I'm not sure a _bet_ is enough to make rainmeter devs accept the potentially compatibility-breaking PR :)

> > The issue I have with this solution is that my LocalPluginLoader will break :)
> 
> plz don't scare me, i prefer living without this feature on breaking that plugin lol, i completely rely on developing my plugins using it.

Well, it would still be possible to make it work, just not automatically, for example with additional manual mechanisms to set whether the plugin uses the new api. Like, adding an options for this. It would be kinda inconvenient if LocalPluginLoader were to crash with some plugins when some special option is not set.

Reply:
>I'm not sure a bet is enough to make rainmeter devs accept the potentially compatibility-breaking PR :)

>For what I know, they are very serious about this, and I agree with them, compatibility is extremely important. If only they were this serious when they were developing their APIs.

I know, but there is no other way to make sure of this, except naming that function with something that doesn't make any sense (`RmGetImageDataFromPluginThatisusedto...`).

>Or I guess I could just make it not work properly with some (or many) plugins, to ensure that there won't be any crashes, which is also not ideal.

That in case we changed the current implementation right? i mean currently, will the plugin (`localPluginLoader`) work fine after the changes we made?

I highly consider naming that function to something that doesn't make sense to avoid any possible issues, and i guess they will understand my point of view, <sub><sup>hopefully</sup></sub>

Reply:
> I know, but there is no other way to make sure of this, except naming that function with something that doesn't make any sense (`RmGetImageDataFromPluginThatisusedto...`).

If only it would be this easy. Because _technically_ someone could already used _any_ name, and it _was valid_ in the old versions.
I really-really-really hate what they did with custom functions :(

> That in case we changed the current implementation right? i mean currently, will the plugin (`localPluginLoader`) work fine after the changes we made?

Well, current version of LocalPluginLoader wouldn't work either way because the API has changed, and it need to be updated for this change.
In this update it will has to have some way to determine whether it should have use new or old api.

Reply:
Don't think about LocalPluginLoader too much. Worst that could happen is that you will have to use some `LocalPluginLoader-WithImages`, or something.

Reply:
>If only it would be this easy. Because technically someone could already used any name, and it was valid in the old versions.

I give up, i rely on my luck that devs are lazy and non of them used a long name.
But i don't think they will accepts such a name anyway, for us we know why it's named like this, but for others they may think it's just a bad naming.

>I really-really-really hate what they did with custom functions :(

Well, what can i say, we just have to live with it :"(

>In this update it will has to have some way to determine whether it should have use new or old api.

hmm, what do you suggest? `RmApiVersion()`? i was about to create a new issue to discuss the possible changes to rainmeter api, like adding other functions and feature that we may use later.

Because in at least 3 years from now, there should be less than 5% of people using old rainmeters (that in case people got a reason to update).

>Don't think about LocalPluginLoader too much. Worst that could happen is that you will have to use some new LocalPluginLoader-WithImages, or something.

Okay but it still hurts a bit, i really liked that you set the plugin name and it just works, section variables, child measure, everything just works.

Reply:
> I give up, i rely on my luck that devs are lazy and non of them used a long name.
> But i don't think they will accepts such a name anyway, for us we know why it's named like this, but for others they may think it's just a bad naming.

If we add new callback, thus breaking the compatibility in case "old rainmeter + new plugin", it will be okay, we can use whatever signature we want.

> hmm, what do you suggest? `RmApiVersion()`? i was about to create a new issue to discuss the possible changes to rainmeter api, like adding other functions and feature that we may use later.

Well, to be honest, I don't know. I think rainmeter devs will dislike any changes unless they are really necessary.
And they will certainly ask why the change is needed, how it will be used and why we can't live without it.
And I really not sure we would be able to push the `get version` function into api.

I guess we can suggest adding something like
```cpp
int32_t RmConfigure(key, value);
int32_t RmSetImageSupport() {return RmConfigure("image-api-1", "true")}
```
This way we could ensure that if anybody even need to extend the API, it can be done without compatibility break.
Though, it would be really good to also have `RmGetConfigValue(key)` to make local plugin loader work, but it will be hard to justify to rainmeter devs.

Reply:
I have an idea of something i still dunno how to do:

```cpp
bool RmExtendApi(feature);

RmExtendApi("LoadImageFromPluginMeasureSupport");
```

And we make this per plugin measure, first, if the plugin didn't find this function then the api must be old, we can log an error.

Second since it's per plugin, any old plugin that didn't have this function will work as expected, because it's just like this feature doesn't exist and the api didn't change.

Also we could extend the api even more if we need to (i guess), because that function will just check if rainmeter supports this feature or not, based on that we can use the feature.

I hope this doesn't seem like a child play, because i don't think every feature request to update the api will be accepted.

Reply:
Your idea is basically identical to mine, you just name you function differently.

Function `int32_t RmConfigure(key, value);` would change the behaviour of rainmeter (just like your `RmExtendApi`), and return a value indicating if the operation was successful. So, we make a version of rainmeter that understands `in-memory-image-transfer`, and we can call `RmConfigure("in-memory-image-transfer", "true")` to extend the API.
Exact key names are discussable, maybe something hierarchical would be better, like `rain.plugin.images`.

On the rainmeter side we support only `in-memory-image-transfer` and if we get anything else, then we return error. Or if we get anything other than `true` as the value, we also return an error.

If someone in the future makes a new feature, then they will add a new value to the list of supported values. If a plugin tries to use it on an old rainmeter version, it gets an error, and it can indicate the correct reason for this error in the rainmeter logs.

There are several issues here:
1. We need reflection for localPluginLoader to work, so I suggested adding `RmGetConfigValue`.
2. We can extend this "soft" api without dll-compatibility break _after_ we already have this soft api.
But right now we don't have this api, and all of the currently avaiable APIs are not suitable for extending.
3. We need to support this in rainmeter side. Every feature that we add through this api will require additional efforts to support it.
So forget about per-plugin features. They are expensive. Rainmeter devs will never go for it.
You personally can create your personal fork of rainmeter with api extension personally for your plugin, create a plugin that is aware of this api and use this plugin universally in both your modded rainmeter and in vanilla rainmeter. This is possible.

Reply:
> 2\. But right now we don't have this api, and all of the currently avaiable APIs are not suitable for extending.

Hmm. I guess this is technically not true.
There is an unofficial `RmGet` function in the API that can be relatively painlessly extended.

Reply:
>Exact key names are discussable, maybe something hierarchical would be better, like `rain.plugin.images`.

Or an enum that contains a list of features :D

>We need reflection for `localPluginLoader` to work, so I suggested adding `RmGetConfigValue`.

What is that function going to do?

>We can extend this "soft" api without dll-compatibility break after we already have this soft api.
>But right now we don't have this api, and all of the currently avaiable APIs are not suitable for extending.

I don't get it.

>So forget about per-plugin features. They are expensive. Rainmeter devs will never go for it.

wait wait wait, what i meant by that is this:

>>On the rainmeter side we support only in-memory-image-transfer and if we get anything else, then we return error. Or if we get anything other than true as the value, we also return an error.

I don't think it's about rm devs, it's about whomever want to add a new feature should actually implement it. We want in-memory image support so we added it, same for other devs, if they want to extend the api they would do that themselves.
It's just should be in a future proof way that can be used by other devs.

>There is an unofficial RmGet function in the API that can be relatively painlessly extended.

But i like the idea behind `RmExtendApi` more :(

Reply:
>Or an enum that contains a list of features :D

Enums are not as easily extendable.

>I don't get it.

When we add a new function to api, the api on dll level breaks, and plugins can't even print a log message what went wrong.
When we change semantics of an api function, plugin can load, get an error and print a convenient log message.

> But i like the idea behind `RmExtendApi` more :(

You can make the actual semantics and most of the syntax the same. It's just a matter of will new plugins load in old versions of not.
`SetConfig = (<set config function ptr>)RmGet(SET_CONFIG)`

Reply:
>`SetConfig = (<set config function ptr>)RmGet(SET_CONFIG)`

A `bool`? if failed then the feature doesn't exist/can't be enabled?

Reply:
You can use `RmGet` to obtain function of the extended api, and then use this function however you want.

On `SetConfig` function: yes, if it failed, then feature can't be enabled.

Reply:
>On `SetConfig` function: yes, if it failed, then feature can't be enabled.

Good, now the question is, how to implement this without rewriting half of rainmeter?
Tbh i still dunno how can this be used (an example of how a plugin may use this).

Reply:
>Good, now the question is, how to implement this without rewriting half of rainmeter?

That's why I said that adding new features is expensive :)
Tomorrow I'll try to look into code and say where exactly we should make changes.

> Tbh i still dunno how can this be used (an example of how a plugin may use this).

I feel like the generic case "plugin tries to enable new extended api" is enough. If an error is returned, it can fallback to another implementation and/or print an explanation into logs.

GetConfig can also be used to finally add version info into rainmeter api, for example.

Reply:
>That's why I said that adding new features is expensive :)

I choosed to suffer since i sold my soul to cpp, so it's fine 😂

>I feel like the generic case "plugin tries to enable new extended api" is enough. If an error is returned, it can fallback to another implementation and/or print an explanation into logs.
>GetConfig can also be used to finally add version info into rainmeter api, for example.

Got it :D